### DISCUSSION TITLE
How can I change options for LGeoJson after some action?

### DISCUSSION THREAD

**Discussion by DoctorChe** [vue-leaflet/vue-leaflet#167] [Q&A] [answered]
I want to download geojson with objects once and then change colors, popups, etc. for these objects after loading various data.
I'm trying to do it as shown below, but it doesn't work.
```html
<template>
  <div class="map" id="container">
    <div id="mapContainer">
      <LMap v-model="zoom" v-model:zoom="zoom" :center="center">
        <LTileLayer :url="url" :attribution="attribution"></LTileLayer>
        <LGeoJson :geojson="MUNICIPALITIES" :options="geojsonOptions"></LGeoJson>
      </LMap>
    </div>
  </div>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
import "leaflet/dist/leaflet.css";
import { LMap, LTileLayer, LGeoJson } from "@vue-leaflet/vue-leaflet";

export default {
  components: { LMap, LTileLayer, LGeoJson },
  name: "Map",
  data() {
    return {
      center: [59.9386, 30.3141],
      zoom: 9,
      url: "https://{s}.basemaps.cartocdn.com/light_all/{z}/{x}/{y}@2x.png",
      attribution:
        '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
      isMunicipalityMode: true,
    };
  },
  computed: {
    ...mapGetters([
      "MUNICIPALITIES",
      "SOME_SCORE",
    ]),
    geojsonOptions() {
      if (this.isMunicipalityMode) {
        return {
          onEachFeature: this.onEachFeatureMunicipality,
        };
      } else {
        return {
          onEachFeature: this.onEachFeatureSomeScore,
        };
      }
    },
    onEachFeatureMunicipality() {
      return (feature, layer) => {
        if (feature.properties && feature.properties.name) {
          layer.bindPopup(`${feature.properties.name}`);
          layer.on("mouseover", () => {
            layer.openPopup();
          });
          layer.on("mouseout", () => {
            layer.closePopup();
          });
        }
      };
    },
    onEachFeatureSomeScore() {
      return (feature, layer) => {
        if (feature.properties && feature.properties.score) {
          layer.bindPopup(`${feature.properties.score}`);
          layer.on("mouseover", () => {
            layer.openPopup();
          });
          layer.on("mouseout", () => {
            layer.closePopup();
          });
        }
      };
    },
  },
  methods: {
    ...mapActions([
      "GET_MUNICIPALITIES_FROM_API",
      "GET_SOME_SCORE_FROM_API",
    ]),
  },
  async mounted() {
    await this.GET_MUNICIPALITIES_FROM_API();
  },
  watch: {
    MUNICIPALITIES() {
      this.GET_SOME_SCORE_FROM_API();
    },
    SOME_SCORE() {
      this.isMunicipalityMode = false;
    },
  },
};
</script>

<style scoped>
#mapContainer {
  height: 100vh;
}
</style>
```

**Comment by mikeu** [comment, marked as answer]
Hi @DoctorChe , that is certainly a decent looking start to what it sounds as if you're trying to accomplish. Could you elaborate more on what exactly isn't working? Is it switching in and out of Municipality Mode where you run into trouble, or putting the data into the popup labels, or having the popups appear on hover instead of on click? Something else?

Here's a pared down version of some of your sample code, showing the popups on some sample geojson, perhaps you can extend it to demonstrate the part that isn't working: https://codesandbox.io/s/vl-geojson-popup-hover-wf8ed?file=/src/App.vue

**Reply by DoctorChe** [reply]
Hi @mikeu.
I have tried to describe my situation here: [https://codesandbox.io/s/vl-geojson-popup-hover-forked-bc9t1?file=/src/App.vue](https://codesandbox.io/s/vl-geojson-popup-hover-forked-bc9t1?file=/src/App.vue).
I have polygons with unique IDs. Next, I load a dataset for these polygons (match them by id). And after loading the data, I want to put the score from dataset in the popup label for each polygon.
And I do not know how can I change data in popups.

**Reply by mikeu** [reply]
Ah, thanks for the clarification @DoctorChe . So the difficulty is in changing the contents of the popups after having bound them. I don't think there would be anything wrong with your approach in principle, except for one thing: at the moment, I don't believe that the `:options` property is reactive.

So, only the initial `onEachFeature` method that was available when the `<l-geo-json>` component was first created ever gets used. Luckily though, Leaflet allows you to use a [function to create popup content](https://leafletjs.com/reference.html#layer-bindpopup)!

Here is another variation on the demo, which binds a popup to each region once when it is created, but whose content is set each time it is opened: https://codesandbox.io/s/vl-geojson-popup-hover-forked-wum9i?file=/src/App.vue

It should let you initially see the "municipality" shape labels, then switch to either score (or even back to municipality mode) as often as you want, with the popups always displaying the correct value for the current mode. Hopefully that gives you enough to get things working in your full application as well.

**Reply by DoctorChe** [reply]
Thanks @mikeu ! This part works great!
But in the next step I need to change the color of the polygons depending on the score value.
I can calculate the color by this way: [https://codesandbox.io/s/vl-geojson-popup-hover-forked-0s1o7?file=/src/App.vue](https://codesandbox.io/s/vl-geojson-popup-hover-forked-0s1o7?file=/src/App.vue)
Is there a way to change the color of polygons after clicking "score 1" or "score 2" button?

**Reply by mikeu** [reply]
For this one, the best way I can think of at the moment is to access the underlying Leaflet GeoJSON object directly to set the new style.

Here's an updated example: https://codesandbox.io/s/vl-geojson-popup-hover-v3-l2del?file=/src/App.vue

It adds a `ref` to the `<l-geo-json>` component, and then when the mode changes it accesses the `.leafletObject` property of it, and calls [`setStyle`](https://leafletjs.com/reference.html#geojson-setstyle) on that. By setting the style to a function that accepts each feature as an argument, you can use the properties of the features to influence their styles. The code isn't the most elegant, it would probably be a good idea to refactor it to have arrays of known modes, scores, styles, labels, etc., but as a toy demo hopefully it gets the idea across.

**Reply by DoctorChe** [reply]
Thank you @mikeu ! It works perfect! Exactly what I need.

**Comment by nielsbom** [comment]
My solution was to force a re-render by giving each LGeoJson a key and changing the key to a random value if I wanted to re-render. [Read more about this technique](https://michaelnthiessen.com/force-re-render/).
Forcing re-renders is more computationally expensive than manually calling `setStyle`. YMMV.
